src/filters/chain.rs

```diff
@@ -124,7 +128,7 @@ impl Filter for FilterChain {
             .try_fold(ctx, |ctx, ((_, filter), histogram)| {
                 Some(ReadContext::with_response(
                     ctx.from,
-                    histogram.observe_closure_duration(|| filter.read(ctx))?,
+                    histogram.observe_closure_duration(|| filter.filter.read(ctx))?,
```

Comment:
Along with changing the type to `FilterInstance` I would change these to call the top level variable `instance`. 
```suggestion
                    histogram.observe_closure_duration(|| instance.filter.read(ctx))?,
```